contracts/asset-proxy/test/uniswap_bridge.ts

```diff
@@ -0,0 +1,384 @@
+                new BigNumber(_opts.amount),
+                // ABI-encoded "from" token address.
+                hexLeftPad(_opts.fromTokenAddress),
+            );
+            return {
+                opts: _opts,
+                result,
+                logs: receipt.logs as any as DecodedLogs,
+                blockTime: await env.web3Wrapper.getBlockTimestampAsync(receipt.blockNumber),
+            };
+        }
+
+        async function getExchangeForTokenAsync(tokenAddress: string): Promise<string> {
+            return testContract.getExchange.callAsync(tokenAddress);
+        }
+
+        it('returns magic bytes on success', async () => {
+            const { result } = await withdrawToAsync();
+            expect(result).to.eq(AssetProxyId.ERC20Bridge);
+        });
+
+        it('just transfers tokens to `to` if the same tokens are in play', async () => {
+            const [[tokenAddress]] = await txHelper.getResultAndReceiptAsync(
+                testContract.createTokenAndExchange,
+                constants.NULL_ADDRESS,
+                '',
+            );
+            const { opts, result, logs } = await withdrawToAsync({
+                fromTokenAddress: tokenAddress,
+                toTokenAddress: tokenAddress,
+            });
+            expect(result).to.eq(AssetProxyId.ERC20Bridge);
+            const transfers = filterLogsToArguments<TokenTransferArgs>(
+                logs,
+                ContractEvents.TokenTransfer,
+            );
+            expect(transfers.length).to.eq(1);
+            expect(transfers[0].token).to.eq(tokenAddress);
+            expect(transfers[0].from).to.eq(testContract.address);
+            expect(transfers[0].to).to.eq(opts.toAddress);
+            expect(transfers[0].amount).to.bignumber.eq(opts.amount);
+        });
+
+        describe('token -> token', () => {
+            it('calls `IUniswapExchange.tokenToTokenTransferInput()', async () => {
+                const { opts, logs, blockTime } = await withdrawToAsync();
+                const exchangeAddress = await getExchangeForTokenAsync(opts.fromTokenAddress);
+                const calls = filterLogsToArguments<TokenToTokenTransferInputArgs>(
+                    logs,
+                    ContractEvents.TokenToTokenTransferInput,
+                );
+                expect(calls.length).to.eq(1);
+                expect(calls[0].exchange).to.eq(exchangeAddress);
+                expect(calls[0].tokensSold).to.bignumber.eq(opts.fromTokenBalance);
+                expect(calls[0].minTokensBought).to.bignumber.eq(0);
+                expect(calls[0].minEthBought).to.bignumber.eq(0);
+                expect(calls[0].deadline).to.bignumber.eq(blockTime);
+                expect(calls[0].recipient).to.eq(opts.toAddress);
+                expect(calls[0].toTokenAddress).to.eq(opts.toTokenAddress);
+            });
+
+            it('sets allowance for "from" token', async () => {
```

Comment:
A test that verifies that the allowance is set on subsequent calls would be nice.